File: processorScripts/mixerChannel.js
class MixerChannel {

    // We pass he superpowered library and sample rate in from the MixerEngine class (whcih passes it on from the AudioWorkletProcessorScript)
    // This also differs slightly from the C++ implemenation of MixerChannel as we ned to handle async loading from the web and not the file system
    constructor(superpoweredInstance, samplerate) {
        this.Superpowered = superpoweredInstance;

        this.channelInputBuffer = new this.Superpowered.Float32Buffer(4096);
        this.channelOutputPreGainBuffer = new this.Superpowered.Float32Buffer(4096);
        this.channelOutputBuffer = new this.Superpowered.Float32Buffer(4096);

        this.peak = 0;
        this.volume = 1;

        this.player = new this.Superpowered.AdvancedAudioPlayer(
            samplerate,
            2,
            2,
            0,
            0.501,
            2,
            false
        );

        this.roll = new this.Superpowered.Roll(
            samplerate,
            samplerate
        );
        this.roll.wet = 1;
        this.roll.enabled = false;

        this.lowpassFilter = new this.Superpowered.Filter(
            this.Superpowered.Filter.Resonant_Lowpass, // The initial filter type
            samplerate // initial sample rate
        );
        this.lowpassFilter.octave = 2;
        this.lowpassFilter.resonance = 0.05;
        this.lowpassFilter.enabled = false;

        this.hipassFilter = new this.Superpowered.Filter(
            this.Superpowered.Filter.Resonant_Highpass, // The initial filter type
            samplerate // initial sample rate
        );
        this.hipassFilter.octave = 2;
        this.hipassFilter.resonance = 0.05;
        this.hipassFilter.enabled = false;
    }

    // Handle async loading of decoded audio buffers from the web
    loadBuffer(buffer, assetDefinition) {
        this.player.openMemory(this.Superpowered.arrayBufferToWASM(buffer), false, false);
        this.assetLoaded = true;
        this.player.originalBPM = assetDefinition.originalBPM;
        this.player.firstBeatMs = assetDefinition.firstBeatMs;
    }

    togglePlay(isMaster) {
        if (!this.playing) {
            if (isMaster) {
                this.player.play()
            } else {
                this.player.playSynchronized();
            }
            this.playing = true;
        } else {
            this.player.pause();
            this.playing = false;
        }
    }

    setVolume(newVolume) {
        this.volume = newVolume;
    }

    getPeak() {
        return this.peak;
    }

    getPitchShift() {
        return this.player.pitchShiftCents/100;
    }

    setPitchShift(newValue) {
        this.player.pitchShiftCents = newValue * 100;
    }

    getIsLoaded() {
        return this.assetLoaded;
    }

    configure(originalBPM, firstBeatMs) {
        this.player.originalBPM = originalBPM;
        this.player.firstBeatMs = firstBeatMs;
        this.player.syncMode = this.Superpowered.AdvancedAudioPlayer.SyncMode_TempoAndBeat;
        this.player.setPosition(this.player.firstBeatMs, false, false);
    }

    getCurrentBPM() {
        return this.player.getCurrentBpm();
    }

    setSyncToBPM(bpm) {
        this.player.syncToBpm = bpm;
        this.roll.bpm = bpm;
    }

    setFilterFrequencyPosition(value) {
        if (value < -20) {
            this.lowpassFilter.frequency = 20020 + value;
            this.lowpassFilter.enabled = true;
            this.hipassFilter.enabled = false;
        } else if (value > 20) {
            this.hipassFilter.frequency = value;
            this.lowpassFilter.enabled = false;
            this.hipassFilter.enabled = true;
        } else {
            this.lowpassFilter.enabled = false;
            this.hipassFilter.enabled = false;
        }
    }

    setRollEnabled(beats) {
        this.roll.beats = beats;
        this.roll.enabled = true;
    }

    setRollDisabled() {
        this.roll.enabled = false;
    }

    process(bufferSize) {

        // Generate player buffers (or silence)
        if (!this.player.processStereo(this.channelOutputBuffer.pointer, false, bufferSize, 1)) {
            for (let n = 0; n < bufferSize; n++) this.channelOutputBuffer.array[n] = 0;
        }

        // Apply roll
        this.roll.process(this.channelOutputBuffer.pointer, this.channelOutputBuffer.pointer, bufferSize);

        // Apply filters
        this.lowpassFilter.process(this.channelOutputBuffer.pointer, this.channelOutputBuffer.pointer, bufferSize);
        this.hipassFilter.process(this.channelOutputBuffer.pointer, this.channelOutputBuffer.pointer, bufferSize);

        // Apply final stage gain
        this.Superpowered.Volume(
            this.channelOutputBuffer.pointer,
            this.channelOutputBuffer.pointer,
            this.volume,
            this.volume,
            bufferSize
        );
    }

    destroy() {
        this.player.destroy();
        this.roll.destroy();
        this.lowpassFilter.destroy();
        this.hipassFilter.destroy();
    }
}

export default MixerChannel

File: cpp/MixerChannel.cpp
//
//  MixerChannel.cpp
//  SuperpoweredUniversalApp (iOS)
//
//  Created by Balázs Kiss on 2021. 11. 23..
//

#include "MixerChannel.hpp"
#include "SuperpoweredSimple.h"

MixerChannel::MixerChannel(MixerChannelAsset asset) {
    this->asset = asset;
    const unsigned int initialSampleRate = 44100;
    sampleRate = initialSampleRate;

    volume.store(0.8);
    peak.store(0);
    isLoaded = false;

    player = new AdvancedAudioPlayer(sampleRate, 0);
    if (asset.fileLength != 0) {
        player->open(asset.filePath, asset.fileOffset, asset.fileLength);
    } else {
        player->open(asset.filePath);
    }

    roll = new Roll(sampleRate);
    roll->wet = 1;
    roll->enabled = false;

    lowPassFilter = new Filter(Filter::FilterType::Resonant_Lowpass, sampleRate);
    lowPassFilter->octave = 2;
    lowPassFilter->resonance = 0.05;
    lowPassFilter->enabled = false;

    highPassFilter = new Filter(Filter::FilterType::Resonant_Highpass, sampleRate);
    highPassFilter->octave = 2;
    highPassFilter->resonance = 0.05;
    highPassFilter->enabled = false;
}

MixerChannel::~MixerChannel() {
    delete player;
    delete lowPassFilter;
    delete highPassFilter;
    delete roll;
}

bool MixerChannel::getIsLoaded() {
    return isLoaded;
}

void MixerChannel::togglePlay(bool isMaster) {
    if (player->isPlaying()) {
        player->pause();
    } else {
        if (isMaster) {
            player->play();
        } else {
            player->playSynchronized();
        }
    }
}

void MixerChannel::setVolume(float newVolume) {
    volume.store(newVolume);
}

float MixerChannel::getPeak() {
    return peak.load();
}

int MixerChannel::getPitchShift() {
    return player->pitchShiftCents / 100;
}

void MixerChannel::setPitchShift(int newValue) {
    player->pitchShiftCents = newValue * 100;
}

void MixerChannel::setSampleRate(unsigned int newSampleRate) {
    this->sampleRate = newSampleRate;
    player->outputSamplerate = newSampleRate;
    roll->samplerate = newSampleRate;
    lowPassFilter->samplerate = newSampleRate;
    highPassFilter->samplerate = newSampleRate;
}

bool MixerChannel::process(float *buffer, unsigned int numberOfFrames, unsigned int sampleRate) {
    if (this->sampleRate != sampleRate) {
        setSampleRate(sampleRate);
    }
    if (player->getLatestEvent() == Superpowered::AdvancedAudioPlayer::PlayerEvent_Opened) {
        isLoaded = true;
    }
    bool hasOutput = player->processStereo(buffer, false, numberOfFrames, 1.0f);

    if (player->syncToBpm != 0) {
        roll->bpm = player->syncToBpm;
    }
    roll->process(buffer, buffer, numberOfFrames);

    lowPassFilter->process(buffer, buffer, numberOfFrames);
    highPassFilter->process(buffer, buffer, numberOfFrames);

    Superpowered::Volume(buffer, buffer, volume.load(), volume.load(), numberOfFrames);
    peak.store(Superpowered::Peak(buffer, numberOfFrames * 2));
    return hasOutput;
}

void MixerChannel::prepareToPlay() {
    player->originalBPM = asset.originalBPM;
    player->firstBeatMs = asset.firstBeatMs;
    player->syncMode = Superpowered::AdvancedAudioPlayer::SyncMode_TempoAndBeat;
    player->setPosition(player->firstBeatMs, false, false);
}

double MixerChannel::getCurrentBPM() {
    return player->getCurrentBpm();
}

void MixerChannel::setSyncToBPM(double bpm) {
    player->syncToBpm = bpm;
}

void MixerChannel::setFilterFrequencyPosition(float newValue) {
    const float maxFrequency = 3000;
    const float minFrequency = 200;

    const float newValueHz = (newValue - 0.5) * (maxFrequency / 0.5);

    if (newValueHz < 0 - minFrequency) {
        highPassFilter->enabled = false;
        lowPassFilter->enabled = true;
        lowPassFilter->frequency = newValueHz + maxFrequency + minFrequency;
    } else if (newValueHz > minFrequency) {
        lowPassFilter->enabled = false;
        highPassFilter->enabled = true;
        highPassFilter->frequency = newValueHz;
    } else {
        lowPassFilter->enabled = false;
        highPassFilter->enabled = false;
    }
}

void MixerChannel::setRollEnabled(float beats) {
    roll->beats = beats;
    roll->enabled = true;
}

void MixerChannel::setRollDisabled() {
    roll->enabled = false;
}
